### ISSUE TITLE
Provide IEnumerable<T> support for Memory<T>

### ISSUE THREAD

**Issue by ahsonkhan** [dotnet/runtime#23950]
## Rationale
Users should be able to pass `Memory<T>` to methods that take in `IEnumerable<T>` similar to array. Currently, this requires calling ToArray() to copy the contents of the `Memory<T>` into an array for it to work, and is unnecessary.

For example (from @eerhardt):
https://github.com/dotnet/corefxlab/blob/master/src/System.Numerics.Tensors/System/Numerics/DenseTensor.cs#L10
```C#
public void SomeMethod(DenseTensor<float> tensor)
{
   Memory<float> memory = tensor.Buffer;
   var value = CreateBatch<T>(..., memory.ToArray(), ...);
}
```

https://github.com/Microsoft/CNTK/blob/master/bindings/csharp/CNTKLibraryManagedDll/ShimApiClasses/ValueShim.cs#L104

```C#
public static Value CreateBatch<T>(NDShape sampleShape, IEnumerable<T> batch, DeviceDescriptor device, bool readOnly = false)
{
    T[] batchAsArray = batch.ToArray();
    return CreateBatch<T>(sampleShape, batchAsArray, 0, batchAsArray.Count(), device, readOnly);
}
```

If `Memory<T>` implemented `IEnumerable<T>`, then the ToArray() call would not be required. However, this would cause applications that reference System.Memory and `Memory<T>` for any scenario to be larger (when compiled AOT). Furthermore, if `Memory<T>` was an `IEnumerable<T>`, it might result in a pit of failure as it could lead to users unintentionally using the enumerator to iterate over the data rather than the more performant indexer on Memory.Span (especially for primitive types like `Memory<byte>`). To discourage unnecessary use of the enumerator but still provide support for the scenarios where you need an IEnumerable, the proposed solution is to add an adapter class and a ToEnumerable extension method on `Memory<T>`.

As an FYI, `Span<T>` cannot implement `IEnumerable<T>` since it is a stack-only, byref type, and casting it to an interface would cause boxing. The compiler will throw and error: `'Span<T>': ref structs cannot implement interfaces`

## Proposed API
```C#
namespace System
{
   public static class MemoryExtensions {
     public static IEnumerable<T> ToEnumerable(this Memory<T> memory);
   }
}
```

## Usage
Taking the above example, it would look as follows:
```C#
public void SomeMethod(DenseTensor<float> tensor)
{
   Memory<float> memory = tensor.Buffer;
   var value = CreateBatch<T>(..., memory.ToEnumerable(), ...);
}
```

## Partial Implementation
```C#
public static IEnumerable<T> ToEnumerable(this Memory<T> memory) => new MemoryEnumerable<T>(memory);

internal class MemoryEnumerable<T> : IEnumerable<T>
{
    ReadOnlyMemory<T> _memory;

    public MemoryEnumerable(ReadOnlyMemory<T> memory) => _memory = memory;

    public IEnumerator<T> GetEnumerator() => new MemoryEnumerator<T>(_memory);

    IEnumerator IEnumerable.GetEnumerator() => new MemoryEnumerator<T>(_memory);
}

internal class MemoryEnumerator<T> : IEnumerator<T>
{
    ReadOnlyMemory<T> _memory;
    int _index;

    public MemoryEnumerator(ReadOnlyMemory<T> memory) => _memory = memory;

    public T Current => _memory.Span[_index];

    object IEnumerator.Current => _memory.Span[_index];

    public void Dispose()
    {
        throw new NotImplementedException();
    }

    public bool MoveNext()
    {
        _index++;
        return _index <_memory.Length;
    }

    public void Reset()
    {
        _index = 0;
    }
}

```

cc @eerhardt, @KrzysztofCwalina, @stephentoub, @jkotas, @terrajobst, @karelz, @ericstj

**Comment by jhudsoncedaron** [comment]
Got here trying to pass a `Memory<char>` to `.All()` linq method. Maybe not the most brilliant way to go about it. Ended up inlining the whole thing.

**Comment by terrajobst** [comment]
Looks [we didn't like](https://github.com/dotnet/corefx/pull/26284#issuecomment-358805150) the API we approved.

@GrabYourPitchforks, could take a look and make a proposal?

**Comment by joshudson** [comment]
I looked at the existing proposal and understand why it wasn't liked. I then called up the actual code of ReadOnlyMemory<T> and studied the Span property.

In my opinion, the only reasonable implementation is to pin `_object` in memory via a fresh `GCHandle` and return a struct iterator from `GetEnumerator()` that implements `IEnumerable` via an actual pointer, and uses `Dispose()` to release the pin on `_object`. This results in no allocations and only one type ladder at loop startup, and only one allocation to pass it to a linq function. I suppose another implementation is possible that basically involves re-implementing the guts of `Span` but it's more dangerous than `GCHandle`.

This is theoretically within my capacity to implement.

**Comment by joshudson** [comment]
So this is my API proposal.

````
partial class System.Memory<T>: IReadOnlyCollection<T> {
    public struct IReadOnlyMemory<T>.Enumerator GetEnumerator();
    int IReadOnlyCollection<T>.Count => Length;
    IEnumerator<T> IEnumerator<T> GetEnumerator() => GetEnumerator();
    IEnumerator IEnumerator.GetEnumerator() => GetEnumerator();
}

partial class System.Memory<T>: IReadOnlyCollection<T> {
    public struct Enumerator GetEnumerator();
    int IReadOnlyCollection<T>.Count => Length;
    IEnumerator<T> IEnumerator<T> GetEnumerator() => GetEnumerator();
    IEnumerator IEnumerator.GetEnumerator() => GetEnumerator();

    public struct Enumerator() : IEnumerable<T> {
        public bool MoveNext();
        public T Current { get ; }
        public void Reset();
    }
}
````

Rationale for `IReadOnlyCollection<T>`: It's just `IEnumerable<T>` + `Count` and it's a convenient flag interface for an enumerable that's not expected to change out from under the caller while the caller retains a reference to it. With this proposed implementation, it's about as good as an Array for being enumerated so there's no downside.

**Comment by stephentoub** [comment]
> the only reasonable implementation is to pin _object in memory via a fresh GCHandle and return a struct iterator from GetEnumerator() that implements IEnumerable via an actual pointer, and uses Dispose() to release the pin on _object

GCHandles are relatively expensive, using one in a struct-based enumerator would make it too easy to accidentally not free one or double-free one, you can't create a pinned handle to arbitrary objects (which would be the case for a `MemoryManager<T>`), and you can't have pointers to arbitrary `T`s (any `T` is possible in `Memory<T>`, not just `unmanaged` types).

If we want to add an enumerator for this, it'll simply access `.Span[i]` on each iteration.  If we want to specialize the implementation for `T[]` in an attempt to optimize for a common case, that could be done as well, assuming performance data proved it out.

e.g. pseudo code with an iterator:
```C#
static IEnumerable<T> Iterate<T>(ReadOnlyMemory<T> memory)
{
    if (MemoryMarshal.TryGetArray(memory, out ArraySegment<T> segment))
    {
        T[] array = segment.Array;
        int end = segment.Offset + segment.Count;
        for (int i = segment.Offset; i < end; i++)
        {
            yield return array[i];
        }
    }
    else
    {
        for (int i = 0; i < memory.Length; i++)
        {
            yield return memory.Span[i];
        }
    }
}
```

**Comment by joshudson** [comment]
> using one in a struct-based enumerator would make it too easy to accidentally not free one

In which case essentially all of the GCHandle sample code is really unsafe.

````
    var gc = GCHandle.Alloc(...);
    try {
        // Stuff here
    } finally {
        gc.Free();
    }
````

Is splittable in .NET Framework (.NET Core can't split it in any way I know). And most of the sample code doesn't even use finally blocks.

I had gotten part-way through the design of an enumerator that uses a string or an array (there's only 3 things it can be) before abandoning it. Maybe that's just better.

**Comment by stephentoub** [comment]
> In which case essentially all of the GCHandle sample code is really unsafe

I don't understand the claim.  The above as a pattern isn't unsafe.

**Comment by joshudson** [comment]
Given the samples as they were, I was under the impression that GCHandle had to work by being found by the GC tracer. As I said, most of the samples don't have finally blocks so the GCHandles don't get freed.

**Comment by stephentoub** [comment]
> most of the samples don't have finally blocks so the GCHandles don't get freed on the inner code taking any exception

Which samples?

**Comment by joshudson** [comment]
Starting from https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.gchandle?view=netcore-3.1

**Comment by stephentoub** [comment]
Yes, that sample should be tweaked, at the very least to move the callback allocation to before the Alloc call.

**Comment by joshudson** [comment]
So what do you think of an implementation that looks like this? I started this path earlier and tossed when I thought of the pointer idea that's going to be more trouble than its worth.

    struct Enumerator {
        private T[] srcArray;
        private string srcString;
        private Utf8String srcUtf8String;
        private int start; // For Reset()
        private int offset;
        private int stop;

        bool MoveNext() { if (offset == stop) return false; /* prevent overflow */ return (++offset != stop); }
        T Current => {
             if (srcArray is object) return srcArray[offset];
             if (srcString is object) return (T)srcString[offset]; /* T is char so cast should disappear at runtime */
             if (srcUtf8String is object) return (T)srcUtf8String[offset];
             throw new InvalidOperationException("...");
        }
    }

**Comment by stephentoub** [comment]
> So what do you think of an implementation that looks like this?

That logic effectively already exists (in a more optimized form) in the `Span` property.  It'd be better to just keep things simple and use `.Span[i]` rather than trying to replicate the logic in Current.

**Comment by GrabYourPitchforks** [comment]
If the goal is to pass a `ROM<T>` to a LINQ method, any struct-based enumerator is going to end up being boxed anyway. I'd honestly rather implement a `MemoryExtensions.ToEnumerable` extension method which returns an `IEnumerable<T>`. The implementation of the `IEnumerable<T>` can be dynamically chosen based on what the `ROM<T>` is actually backed by: an array, a string, or something else. The method would be documented as "yup, this allocates the enumerator, but the scenario was that you're going to pass it to LINQ anyway, so... 🤷‍♀️."

```cs
public static class MemoryExtensions
{
    public static IEnumerable<T> ToEnumerable(this ReadOnlyMemory<T> memory);
}
```

**Comment by joshudson** [comment]
@GrabYourPitchforks : That's where we started at the top of the issue and the API team rejected it. So I acted on the reason that was found, that is it's a trap if you enumerate it directly.

**Comment by GrabYourPitchforks** [comment]
@joshudson Not quite. The implementation I suggested is different and should avoid the performance concerns. I suggested deconstructing the `Memory<T>` instance _at the call to `ToEnumerable()`_, which means that the deconstruction cost occurs upfront when the enumerator is created - not on each call to `MoveNext`. The public API would look identical to what has already been approved.

Things get complicated if you're backed by a `MemoryManager<T>` instead of a `string` or a `T[]`, and enumerating an instance of that would still incur some overhead. But that shouldn't represent the majority case.

**Comment by terrajobst** [comment]
Sounds good.

**Comment by Shpendicus** [comment]
Hey guys, 

so we can do: foreach on spans, and AFAIK .NET makes a promise to use Duck-Typing for such constructs, like GetPinnableReference() to be used in fixed(....) but why is the compiler not smart enough (.NET 6.0 ) to infer that Span<T> is duck-typed to an IEnumerable<T> without implementing it

Cause when an extension method takes a T, where T : IEnumerable>T>, it could use internal Duck-Typing to infer, the type actually does the same as it being an IEnumerable<T>.

Would be really nice if .NET could make ducktyping a thing, i mean coding all these LINQ things will be quite ugly :D

**Comment by jkotas** [comment]
> why is the compiler not smart enough (.NET 6.0 ) to infer that Span is duck-typed to an IEnumerable without implementing it

C# compiler does infer that in .NET 6. `foreach` works on Span for several years. More discussion on this is captured in https://github.com/dotnet/csharplang/blob/main/meetings/2017/LDM-2017-12-04.md#foreach-for-span-and-readonlyspan . The discussion on duck-typing is more appropriate for https://github.com/dotnet/csharplang repo.

**Comment by Shpendicus** [comment]
> C# compiler does infer that in .NET 6. `foreach` works on Span for several years.

I already said that, good Sir, namely below: 
@jkotas 
> so we can do: foreach on spans

This is a good thing, its just, we cannot make use of the Enumerable-LINQ-Extension methods sadly, cause the Compiler still strictly wants the callee of these methods to be of type IEnumerable<T>.

That is my Dilemma

**Comment by stephentoub** [comment]
The C# compiler can't fabricate a wrapper IEnumerable for the same reason Span doesn't implement IEnumerable: it's a ref struct that can't be boxed or otherwise copied to the heap.

**Comment by Shpendicus** [comment]
@Joe4evr i think they are on a really good track, I also saw a nuget package called: **NoAlloc** maybe you can try that one aswell 🗡️

**Comment by eiriktsarpalis** [comment]
Marking as ready-for-review in light of the discussion in https://github.com/dotnet/runtime/issues/77209#issuecomment-1339857514. It has been suggested that `ReadOnlyMemory<T>` implementing `IEnumerable<T>` directly is desirable, since it would avoid the need for a dedicated `IEqualityComparer<ReadOnlyMemory<T>> CreateMemoryComparer<T>(IEqualityComparer<T>)` method and we can specialize `IEqualityComparer<TEnumerable> CreateEnumerableComparer<T>(IEqualityComparer<T>) where TEnumerable : IEnumerable<T>` instead.

**Comment by bartonjs** [comment]
This came up in API Review, and the main question that we had was "why is `MemoryMarshal.ToEnumerable(ReadOnlyMemory<T>)` not good enough?".  We didn't seem to see any discussion related to that existing member.

**Comment by jhudsoncedaron** [comment]
@bartonjs : I just now looked into it. It looks like this issue is an example of discovery fail. Oof. I would never think to look in namespace interopservices for something like that.

It's also a pretty terrible implementation, performance-wise. I left behind a much faster one in a comment above.

**Comment by terrajobst** [comment]
I think this depends on what you want to achieve. If you just want to enumerate the contents in a foreach loop, then you shouldn't use this API at all. You should get the span and foreach the span which the compiler will basically emit as for-loop over the span and the JIT will probably be able to hoist bound checks outside the loop, which is as performant as you can get.

But if you need to pass the memory to an API that wants an `IEnumerable<T>` then you need to allocate a class. You could try to play tricks to avoid the allocation of the enumerator but that isn't necessarily safe if you want to allow multiple enumerations from the same instance. And I don't think the implementation can avoid requesting the span for each element, which isn't that cheap either. But I think all this overhead comes from the fact that the caller wants to go through `IEnumerable<T>`.

**Comment by jhudsoncedaron** [comment]
"And I don't think the implementation can avoid requesting the span for each element"

Yes it can. Since it is inside System.Private.CoreLib it can unwrap the Memory<T>, which is what I posted. Oh well.

**Comment by terrajobst** [comment]
I can't seem to find your comment. Could you link it?

**Comment by DaZombieKiller** [comment]
The APIs for unwrapping a `Memory<T>` are actually public (`MemoryMarshal.TryGetArray`, `MemoryMarshal.TryGetString` and `MemoryMarshal.TryGetMemoryManager`) so even if the method wasn't in the BCL it could do that.

The problem is when the `Memory<T>` isn't backed by a string or an array but rather a `MemoryManager`. The implementation could be made more efficient by using `Pin`, but as mentioned in API review it would most likely require a finalizer which would be non-ideal.

**Comment by terrajobst** [comment]
I see, you're suggesting an implementation for `ToEnumerable()` that can special case specific underlying types for a span (such as arrays and strings) and provide an implementation that wouldn't have to get the span for each element. I think that's fair and seems quite doable. For short spans, this might make it worse though, but we could measure that. For native memory or custom memory managers it would probably continue to be somewhat inefficient if we want some amount of safety. However, I can buy that these scenarios are more rare in practice as they most likely deal with byte buffers which one generally doesn't want to enumerate but parse.

**Comment by DaZombieKiller** [comment]
We're essentially entering the space of "niche within niche" here, but if you want to allow as many cases to be efficient as possible, it may make sense to add a new `virtual` method to `MemoryManager`:

```cs
namespace System.Buffers;

public abstract partial class MemoryManager<T>
{
    public virtual IEnumerable<T> ToEnumerable()
    {
        // Default implementation using 'yield return'
    }
}
```

An example of the kind of code this might benefit from this is in the community toolkit. [There are APIs in the toolkit that offer the equivalent of `MemoryMarshal.Cast` for `Memory<T>` instances](https://github.com/CommunityToolkit/dotnet/blob/e071ed22b379f904820475db827be3ad04d3e96e/src/CommunityToolkit.HighPerformance/Extensions/ReadOnlyMemoryExtensions.cs#L82-L132), which hand back `MemoryManager<T>`-backed `Memory<T>`s that provide `MemoryMarshal.Cast`-ed `Span<T>`s. Having that extra virtual would allow those `MemoryManager`s (and any others like them) to avoid the performance penalty associated with the default `yield return` implementation.

Is that worth it for such a niche use case though? I'm skeptical.

**Comment by stephentoub** [comment]
> It's also a pretty terrible implementation, performance-wise.

https://github.com/dotnet/runtime/pull/89274 I put up earlier fixes that.

**Comment by jhudsoncedaron** [comment]
@stephentoub  Ah good that problem's going away.

**Comment by ionoy** [comment]
Is there a reason this wasn't exposed as an extension method?

**Comment by tannergooding** [comment]
This isn't something we're planning on providing today. `MemoryMarshal.ToEnumerable` *is* good enough and devs can author their own extension methods if really desired.
